[SHK-64] FeedProduct 등록, 조회 기능 구현

Kim-Changgyu opened the pull request:
### ⛏ 작업 사항
- FeedProduct 엔티티 및 FeedProductRepository 구현
- Feed 엔티티에 필드 추가
  - 상품 태그 (즉, 상품 ID값을 가지는 리스트)
  - 해당 필드는 DB에 저장하지 않도록 @Transient
- DTO에 Product 식별자 리스트 추가
- 상품 태그 등록, 조회 관련 단위 테스트 코드 추가

### 📝 작업 요약
- 피드의 상품 태그 등록, 조회 기능 구현

### 💡 관련 이슈
- Resolved : [SHK-166], [SHK-179]


[SHK-166]: https://shoekream.atlassian.net/browse/SHK-166?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SHK-179]: https://shoekream.atlassian.net/browse/SHK-179?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Kim-Changgyu committed:
Feat : [SHK-166] 피드 상품 태그 등록 기능 추가

- 피드 등록 플로우에서 상품 태그를 검색해서 리스트로 넘긴다는 가정

Kim-Changgyu committed:
Test : 피드 상품 태그 등록 기능 단위 테스트

Kim-Changgyu committed:
Feat : [SHK-179] 피드 조회 응답에 상품 태그 추가

- 테이블의 상품 식별자 리스트를 받도록 구현

Kim-Changgyu committed:
Test : 피드 상품 태그 포함 조회 기능 단위 테스트

BnDC commented on src/main/java/com/prgrms/kream/common/mapper/StyleMapper.java:
```diff
@@ -61,6 +64,15 @@ public static Feed toFeed(RegisterFeedServiceRequest registerFeedServiceRequest)
 				.build();
 	}
 
+	public static List<FeedProduct> toFeedProduct(Long feedId, List<Long> productIds) {
+		return productIds.stream()
+				.map(productId -> FeedProduct.builder()
+						.feedId(feedId)
+						.productId(productId)
+						.build())
+				.toList();
+	}
```
`toFeedProductList` 어떠신가요?

Kim-Changgyu replied:
헉 이 부분을 놓쳤네요 !
제가 작성한 코드의 대부분이 List인 경우에 s만 붙이도록 규칙을 정해두어서 toFeedProducts 로 변경하도록 하겠습니다 :)

BnDC commented on src/test/java/com/prgrms/kream/domain/style/service/StyleServiceTest.java:
```diff
@@ -65,15 +70,31 @@ class StyleServiceTest {
 	void testRegister() {
 		when(feedRepository.save(any())).thenReturn(FEED);
 		when(feedTagRepository.saveAll(any())).thenReturn(FEED_TAGS.stream().toList());
+		when(feedProductRepository.saveAll(any())).thenReturn(FEED_PRODUCTS);
```
단순히 `any()` 보다는 `any(Entity.class)` 로 만드는 건 어떨까요?

BnDC commented on src/test/java/com/prgrms/kream/domain/style/service/StyleServiceTest.java:
```diff
@@ -65,15 +70,31 @@ class StyleServiceTest {
+
 		RegisterFeedServiceResponse feedResponse = styleService.register(getRegisterFeedServiceRequest());
 
 		assertThat(feedResponse.id()).isEqualTo(1L);
```
결과값 검증과 함께 `verify()` 로 실제 Mocking된 함수가 호출되는지도 검증해보면 좋을 거 같습니다

BnDC reviewed (approved):
구현하느라 고생하셨습니다
![StandingOvationGIF](https://user-images.githubusercontent.com/86050295/215254818-bba39c4f-bcf9-4bc1-b003-e755f0c184cd.gif)

Kim-Changgyu committed:
Feat : 피드 조회(인기순) 커서 기반 페이지네이선 적용

- [페이지네이션 관련 참고자료](https://velog.io/@minsangk/%EC%BB%A4%EC%84%9C-%EA%B8%B0%EB%B0%98-%ED%8E%98%EC%9D%B4%EC%A7%80%EB%84%A4%EC%9D%B4%EC%85%98-Cursor-based-Pagination-%EA%B5%AC%ED%98%84%ED%95%98%EA%B8%B0#%EC%BC%80%EC%9D%B4%EC%8A%A4-price-desc-id-desc-%EC%BB%A4%EC%8A%A4%ED%85%80-cursor-%EC%83%9D%EC%84%B1) 토대로 개발
- WHERE절 조건이 간단하므로 성능에 큰 부담이 되진 않을 것으로 추측 (추후 성능 측정 예정)

boompatron commented on src/main/java/com/prgrms/kream/domain/style/dto/response/GetFeedFacadeResponse.java:
```diff
@@ -8,6 +8,7 @@ public record GetFeedFacadeResponse(
 		Long authorId,
 		String content,
 		Long likes,
+		List<Long> products,
```
필드 명이 products 라 되어있는데 productId가 아닌 이유가 있을까요??

Kim-Changgyu replied:
현재는 ID만 가지만 추후에 상품명 같은 데이터를 포함시킬 예정이라 필드명을 미리 세팅해두었습니다 !

boompatron reviewed (approved):
수고하셨습니다!! 전체적으로 필드를 하나 추가하신 것, 페이징 처리를 위주로 작업을 하신 것 같네요!! 질문 1개 남겼습니다!!

Kim-Changgyu committed:
Refactor : [SHK-64] 피드백 반영

- 단위 테스트 각 테스트 메서드에 verify 보완
- Mapper 메서드명 수정 (리스트이기 때문에 toFeedProduct -> toFeedProducts)
